### PR TITLE
Apply makeNotAllowedError to empty Git repos

### PR DESCRIPTION
**Pull request by edolstra**
## Motivation

Without this, the `makeNotAllowedError` provided by the Git fetcher doesn't work for empty repos, so you don't get its helpful messages (like "access to path 'flake.nix' is forbidden because it is not under Git control") if nothing has been `git add`ed to the repo yet.

This requires `AllowListSourceAccessor` to support allowed paths in addition to allowed prefixes, since we need to allow access to the root of the Git repo but not any of its children.

## Context

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr/eval.cc [hunk 1]
```diff
@@ -273,7 +273,7 @@ EvalState::EvalState(
 
             /* Apply access control if needed. */
             if (settings.restrictEval || settings.pureEval)
-                accessor = AllowListSourceAccessor::create(accessor, {},
+                accessor = AllowListSourceAccessor::create(accessor, {}, {},
```
Comment (Mic92) [inline src/libexpr/eval.cc:276]: Is it possible to add a test case for empty repositories?

Comment (edolstra) [reply src/libexpr/eval.cc:276]: Added a test.
